### ISSUE TITLE
Wrong "depth relationship" of 3D isosurfaces

### ISSUE THREAD

**Issue by brisvag** [napari/napari#2418]
## 🐛 Bug

Image layers in 3D are drawn with a simple bottom-to-top order, which is not what one would expect from isosurfaces.
For example:
![](https://i.ibb.co/1JDCN9M/Peek-2021-03-15-14-43.gif)

(I think I remember seeing similar behaviour with other visualizations, though I can't recall exactly what it was. I couldn't find an issue about this, but maybe I'm looking for the wrong keywords.)

## To Reproduce

A simple example:
```python
v = napari.Viewer(ndisplay=3)
v.add_image(np.arange(1000).reshape(10, 10, 10), iso_threshold=0.7, rendering='iso')
v.add_image(np.arange(1000, 0, -1).reshape(10, 10, 10), iso_threshold=0.7, rendering='iso')
```
Then move it around.

## Expected behavior

The "depth relationship" of the isosurfaces should be determined based on the physical distance from the camera, rather than simply be based on layer order. I'm not sure if layer order should have any meaning here, of if I'm missing some important point.

## Environment

```yaml
napari: 0.4.6
Platform: Linux-5.11.4-arch1-1-x86_64-with-glibc2.33
System: Arch Linux
Python: 3.9.2 (default, Feb 20 2021, 18:40:11) [GCC 10.2.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.20.1
SciPy: 1.6.1
Dask: 2020.12.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 460.56
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0
- screen 2: resolution 1920x1200, scale 1.0

Plugins:
- console: 0.0.3
- mrcfile-reader: 0.1.2
- napari-em-reader: 0.1.0
- ndtiffs: 0.1.1
- svg: 0.1.4
```

**Comment by sofroniewn** [comment]
@brisvag can you try changing the camera `fov` (right now a private method deep in vispy) and let us know what happens. See this line https://github.com/napari/napari/issues/2320#issuecomment-789250845 and the discussion around it

**Comment by brisvag** [comment]
Pretty cool, the fov change! But no, this seems to be unaffected by the fov.

**Comment by sofroniewn** [comment]
ah ok, then this is really an issue with having the data in multiple layers - try switching to `additive` blending instead of translucent. That may have other slightly undesirable properties. Can you share the data files for your example above - it's a little harder for me to intuitively grasp the expected behavior of the toy example

**Comment by brisvag** [comment]
I've got a better toy example:
```python
import skimage
import napari

v = napari.Viewer(ndisplay=3)

blobs = skimage.data.binary_blobs(length=128, blob_size_fraction=0.2, volume_fraction=0.2, n_dim=3)
blobs2 = skimage.data.binary_blobs(length=128, blob_size_fraction=0.2, volume_fraction=0.2, n_dim=3)

v.add_image(blobs, rendering='iso', colormap='green')
v.add_image(blobs2, rendering='iso', colormap='magenta')
```

This should be even more obvious. The green layer is always behind the magenta layer. Additive doesn't quite solve the problem, because it introduces transparency. In some cases, I just want to only see the volume that's closer to me!

Right now it's as if each layer is rendered as part of a separate 3d space, and then the rendered image is superimposed to the previous layer. Instead, I would expect the two layers to belog to the same 3d space, and thus cover each other where appropriate.

**Comment by sofroniewn** [comment]
Thanks for the example I'll play with this more when I get the chance

> Right now it's as if each layer is rendered as part of a separate 3d space, and then the rendered image is superimposed to the previous layer. Instead, I would expect the two layers to belog to the same 3d space, and thus cover each other where appropriate.

I guess that is how our layer ordering works now, I think @jni and I have talked about this before in the context of labels and images that get overlaid. I guess in this context what you're proposing makes sense, but I wonder if that is always true - would people ever want to do have one layer entirely on top of the other layer.

There is an `order` parameter on the visual, I can try manually overwriting it in this case and it may fix things ... we'll see

**Comment by jni** [comment]
As I understand the problem yes, our layers always get drawn one behind the other, and I think it's a limitation of the architecture of vispy visuals — they get run in the shader in sequence, so the last one to run is the one you see — there's no concept of a spatial relationship as the shader gets executed. It would be good if @djhoese or @tlambert03 could confirm this.

@sofroniewn I think this is the same reason your stick meshes are always behind your sphere visuals in https://github.com/vispy/vispy/pull/2008#issuecomment-798829246

I don't think it's an easy fix unfortunately. :sob:

**Comment by djhoese** [comment]
My go-to person for this is usually @rougier. OpenGL artifact stuff is really not something I understand well enough to guess beyond what's already been said here. @rougier any ideas on if this looks like something caused by drawing order? Any ideas on a solution?

**Comment by rougier** [comment]
It looks like the depth buffer is not written by each volume such that the rendering becomes order dependent for multiple volumes. The faulty line is [here](https://github.com/vispy/vispy/blob/master/vispy/visuals/volume.py#L263). You can also have a look at https://www.labri.fr/perso/nrougier/python-opengl/#spheres to see how this works.

One quick and dirty solution for multiple separable volumes would be to render them in the order of their center. If not really separable, you'll need to fix the shader such as to write the depth buffer (and make sure depth buffer is enabled). Note that this will slow down rendering.

**Comment by sofroniewn** [comment]
> You can also have a look at https://www.labri.fr/perso/nrougier/python-opengl/#spheres to see how this works.

Wow @rougier thanks so much for that link! I can't believe I havn't looked at your online book before, that is incredibly helpful! cc @alisterburt @kevinyamauchi who were recently asking about some OpenGL stuff, you might want to look in here!

**Comment by brisvag** [comment]
As we said in meeting, it might be a good idea to keep #2138 in mind as well, as it's probably related in some way.

**Comment by brisvag** [comment]
This was closed by #3181! :) :tada:
